### PR TITLE
[crimson]:seastar-osd perfcounters support

### PR DESCRIPTION
**Pull request by liu-chunmei**
move m_lock out from perfcounters.cc to upper proxy,then seastar osd can reuse perfcounters.cc
create shard service for perfcounters, each seastar thread has its own perfcounterscollection and
perfcunters m_loggers.

Signed-off-by: chunmei Liu <chunmei.liu@intel.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by liu-chunmei** [comment]
@tchaikov @cyx1231st, please help review the PR.

File: src/crimson/common/perf_counters_service.cc [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#include "perf_counters_service.h"
+namespace ceph::common {
+PerfCountersService::PerfCountersService()
+{
+  if (seastar::engine().cpu_id() != 0) {
+    return;
+  }
+  pfCollection.reset(new PerfCountersCollection);
```
Comment (tchaikov) [inline src/crimson/common/perf_counters_service.cc]: please refer https://github.com/ceph/ceph/blob/master/CodingStyle for the naming convention. in general, we don't use Hungarian notation in C++ code. `perf_collection` or `perf_coll` might be better alternative.

File: src/crimson/common/perf_counters_service.cc [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+}
+PerfCountersService::~PerfCountersService()
+{
+  pfCollection.reset();
```
Comment (tchaikov) [inline src/crimson/common/perf_counters_service.cc]: `pfCollection` is already a smart pointer. so no need to reset it explicitly.

File: src/pybind/mgr/dashboard/frontend/package-lock.json [hunk 3]
```diff
@@ -702,8 +702,7 @@
     "abbrev": {
       "version": "1.0.9",
       "resolved": "https://registry.npmjs.org/abbrev/-/abbrev-1.0.9.tgz",
-      "integrity": "sha1-kbR5JYinc4wl813W9jdSovh3YTU=",
```
Comment (tchaikov) [inline src/pybind/mgr/dashboard/frontend/package-lock.json]: please drop unrelated change.

File: src/test/crimson/test_perfcounters.cc [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#include <pthread.h>
+#include <iostream>
+#include "common/Formatter.h"
+#include <seastar/core/app-template.hh>
+#include <seastar/core/sharded.hh>
+#include "common/perf_counters.h"
+#include "crimson/common/perf_counters_service.h"
```
Comment (tchaikov) [inline src/test/crimson/test_perfcounters.cc]: nit, add a empty line after the last `#include`

File: src/test/crimson/test_perfcounters.cc [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+using PerfCountersService = ceph::common::PerfCountersService;
```
Comment (tchaikov) [inline src/test/crimson/test_perfcounters.cc]: `using ceph::common::PerfCountersService;` would suffice.

File: src/test/crimson/test_perfcounters.cc [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+static seastar::future<> test_perfcounters(){
+  return ceph::common::sharded_pfService().start().then([] {
+    return ceph::common::sharded_pfService().invoke_on(0,&PerfCountersService::start);
+  }).then([] {
+    return ceph::common::sharded_pfService().invoke_on_all([] (auto& s){
+         PerfCounters *perf_logger;
+         char name[128];
+         sprintf(name, "seastar-osd::thread-%d", pthread_self());
```
Comment (tchaikov) [inline src/test/crimson/test_perfcounters.cc]: printing the shard id makes more sense in this context.

File: src/test/crimson/test_perfcounters.cc [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  perftest_count,
+  perftest_last,
+};
+
+static seastar::future<> test_perfcounters(){
+  return ceph::common::sharded_pfService().start().then([] {
+    return ceph::common::sharded_pfService().invoke_on(0,&PerfCountersService::start);
+  }).then([] {
+    return ceph::common::sharded_pfService().invoke_on_all([] (auto& s){
+         PerfCounters *perf_logger;
```
Comment (tchaikov) [inline src/test/crimson/test_perfcounters.cc]: wrong indent.

File: src/test/crimson/test_perfcounters.cc (hunk 6 above)
Comment (tchaikov) [reply src/test/crimson/test_perfcounters.cc]: nit, could use libfmt for fprint.

File: src/test/crimson/test_perfcounters.cc [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+         perf_logger = plb.create_perf_counters();
+         perf_logger->inc(perftest_count);
+         s.pfCollection->add(perf_logger);
+       });
+     }).then([] { 
+         return ceph::common::sharded_pfService().invoke_on_all([] (auto& s){
+         ostringstream ostr;
+         Formatter* f = Formatter::create("json-pretty", "json-pretty", "json-pretty");
+         s.pfCollection->dump_formatted(f, false);
+         ostr << "dump perfcounters";
```
Comment (tchaikov) [inline src/test/crimson/test_perfcounters.cc]: might want to verify the "perftest_count" and fail the test it's not expected?

File: src/common/perf_counter_proxy.h [hunk 9]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "common/perf_counters.h"
+#include "common/Mutex.h"
+
+class CephContext;
+/* PerfcounterCollectionProxy hold the lock for PerfCounterCollection */
+class PerfCountersCollectionProxy
```
Comment (tchaikov) [inline src/common/perf_counter_proxy.h]: regarding to the naming: how about renaming existing `PerfCountersCollection` to `PerfCountersCollectionImpl` and just extract the locking into `PerfCountersCollection` ?

File: src/common/perf_counter_proxy.h [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+{
+  CephContext *m_cct;
+
+  /** Protects m_loggers */
```
Comment (tchaikov) [inline src/common/perf_counter_proxy.h]: could instead put
```c++
 // Protects PfCollection->m_loggers
```

File: src/common/perf_counter_proxy.h [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+                      const std::string &logger = "",
+                      const std::string &counter = "");
+  void dump_formatted_histograms(ceph::Formatter *f, bool schema,
+                                 const std::string &logger = "",
+                                 const std::string &counter = "");
+
+  void with_counters(std::function<void(const PerfCountersCollection::CounterMap &)>) const;
+
+  friend class PerfCountersCollectionTest;
+};
```
Comment (tchaikov) [inline src/common/perf_counter_proxy.h]: add an empty line after `};`

**Comment by tchaikov** [comment]
@liu-chunmei have you put some thoughts on integrating ceph's PerfCounter system with `seastar::metrics` ? as we will be also interested in seastar's internal metrics in future. how shall we expose seastar's metrics with existing PerfCounter infrastructure?

i know, this brings us to a more general question, how shall we integrate seastar's logging and configuration systems with our logging and config systems... i admit i neglect this problem before and i will also also think about it..

**Comment by liu-chunmei** [comment]
@tchaikov I will consider merger perfcounters and seastar metrics at the following steps.

File: src/common/perf_counter_proxy.h (hunk 9 above)
Comment (liu-chunmei) [reply src/common/perf_counter_proxy.h]: sure, will do

**Comment by liu-chunmei** [comment]
I will update the PR.
